File: backend/app/models/season.py
"""Crop season -- the organising unit for every other record."""

from __future__ import annotations

from decimal import Decimal
from typing import TYPE_CHECKING

from sqlalchemy import BigInteger, CheckConstraint, Index, Numeric, String, Text, text
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.db.base import Base, SyncMixin
from app.models.enums import SeasonStatus, sql_in_list

if TYPE_CHECKING:  # forward refs; resolved at mapper-configuration time
    from app.models.diary import DiaryEntry
    from app.models.finance import Expense, Revenue
    from app.models.inventory import StockTransaction


class Season(Base, SyncMixin):
    """e.g. "Vụ Đông Xuân 2026" -- rice, 5 sào, Dec 2026 to Apr 2027."""

    __tablename__ = "seasons"

    __extra_table_args__ = (
        CheckConstraint(
            f"status IN ({sql_in_list(SeasonStatus.values())})",
            name="status_valid",
        ),
        # Enforced in three layers -- table, Pydantic, and the mobile form.
        # A bad range silently breaks every report that filters by the season
        # window, and a silent report bug is far more expensive to find than a
        # loud validation error.
        CheckConstraint(
            "end_date IS NULL OR end_date >= start_date",
            name="date_range_valid",
        ),
        CheckConstraint("area_size IS NULL OR area_size >= 0", name="area_non_negative"),
        CheckConstraint("length(name) BETWEEN 1 AND 120", name="name_length"),
        CheckConstraint("length(crop_type) BETWEEN 1 AND 80", name="crop_type_length"),
        Index(
            "ix_seasons_household_start",
            "household_id",
            text("start_date DESC"),
            postgresql_where=text("deleted_at IS NULL"),
        ),
    )

    name: Mapped[str] = mapped_column(Text, nullable=False)
    crop_type: Mapped[str] = mapped_column(Text, nullable=False)
    area_size: Mapped[Decimal | None] = mapped_column(Numeric(10, 3), nullable=True)
    area_unit: Mapped[str] = mapped_column(String(16), nullable=False, server_default=text("'sao'"))

    start_date: Mapped[int] = mapped_column(BigInteger, nullable=False)
    # NULL means the season is still running -- the farmer has not yet decided
    # when it ends, which is normal mid-season.
    end_date: Mapped[int | None] = mapped_column(BigInteger, nullable=True)

    status: Mapped[str] = mapped_column(
        String(16), nullable=False, server_default=text("'active'")
    )
    note: Mapped[str | None] = mapped_column(Text, nullable=True)

    diary_entries: Mapped[list["DiaryEntry"]] = relationship(
        back_populates="season", passive_deletes=True
    )
    expenses: Mapped[list["Expense"]] = relationship(
        back_populates="season", passive_deletes=True
    )
    revenues: Mapped[list["Revenue"]] = relationship(
        back_populates="season", passive_deletes=True
    )
    stock_transactions: Mapped[list["StockTransaction"]] = relationship(
        back_populates="season", passive_deletes=True
    )

File: mobile/src/db/models/Season.ts
import {Model, Q, type Query} from '@nozbe/watermelondb';
import type {Associations} from '@nozbe/watermelondb/Model';
import {children, date, field, lazy, readonly, text} from '@nozbe/watermelondb/decorators';

import type {SeasonStatus} from '../enums';
import type DiaryEntry from './DiaryEntry';
import type Expense from './Expense';
import type Revenue from './Revenue';
import type StockTransaction from './StockTransaction';

/** Mùa vụ — the organising unit for every other record. */
export default class Season extends Model {
  static table = 'seasons';

  static associations: Associations = {
    diary_entries: {type: 'has_many', foreignKey: 'season_id'},
    stock_transactions: {type: 'has_many', foreignKey: 'season_id'},
    expenses: {type: 'has_many', foreignKey: 'season_id'},
    revenues: {type: 'has_many', foreignKey: 'season_id'},
  };

  @text('name') name!: string;
  @text('crop_type') cropType!: string;
  @field('area_size') areaSize!: number | null;
  @text('area_unit') areaUnit!: string;
  @date('start_date') startDate!: Date;
  /** null = the season is still running; the farmer has not decided when it ends. */
  @date('end_date') endDate!: Date | null;
  @text('status') status!: SeasonStatus;
  @text('note') note!: string | null;

  @readonly @date('created_at') createdAt!: Date;
  @date('updated_at') updatedAt!: Date;

  @children('diary_entries') diaryEntries!: Query<DiaryEntry>;
  @children('stock_transactions') stockTransactions!: Query<StockTransaction>;
  @children('expenses') expenses!: Query<Expense>;
  @children('revenues') revenues!: Query<Revenue>;

  /** Live consumption for this season — drives the supply-consumption chart. */
  @lazy consumption = this.stockTransactions.extend(Q.where('txn_type', 'out'));
}
